File: packages/jsii-reflect/lib/docs.ts
import * as jsii from '@jsii/spec';
import { Stability } from '@jsii/spec';

import { TypeSystem } from './type-system';

export interface Documentable {
  docs: Docs;
}

export class Docs {
  public readonly docs: jsii.Docs;

  public constructor(
    public readonly system: TypeSystem,
    public readonly target: Documentable,
    spec: jsii.Docs,
    private readonly parentDocs?: Docs,
  ) {
    this.docs = spec ?? {};
  }

  /**
   * Returns docstring of summary and remarks
   */
  public toString() {
    return [this.docs.summary, this.docs.remarks]
      .filter((txt) => !!txt)
      .join('\n\n');
  }

  public get subclassable(): boolean {
    return !!this.docs.subclassable;
  }

  /**
   * Return the reason for deprecation of this type
   */
  public get deprecationReason(): string | undefined {
    if (this.docs.deprecated !== undefined) {
      return this.docs.deprecated;
    }
    if (this.parentDocs) {
      return this.parentDocs.deprecationReason;
    }
    return undefined;
  }

  /**
   * Return whether this type is deprecated
   */
  public get deprecated(): boolean {
    return this.deprecationReason !== undefined;
  }

  /**
   * Return the stability of this type
   */
  public get stability(): jsii.Stability | undefined {
    return lowestStability(this.docs.stability, this.parentDocs?.stability);
  }

  /**
   * Return any custom tags on this type
   */
  public customTag(tag: string): string | undefined {
    return this.docs.custom?.[tag];
  }

  /**
   * Return summary of this type
   */
  public get summary(): string {
    return this.docs.summary ?? '';
  }

  /**
   * Return remarks for this type
   */
  public get remarks(): string {
    return this.docs.remarks ?? '';
  }

  /**
   * Return examples for this type
   */
  public get example(): string {
    return this.docs.example ?? '';
  }

  /**
   * Return documentation links for this type
   */
  public get link(): string {
    return this.docs.see ?? '';
  }

  /**
   * Returns the return type
   */
  public get returns(): string {
    return this.docs.returns ?? '';
  }

  /**
   * Returns the default value
   */
  public get default(): string {
    return this.docs.default ?? '';
  }
}

const stabilityPrecedence = {
  [Stability.Deprecated]: 0,
  [Stability.Experimental]: 1,
  [Stability.External]: 2,
  [Stability.Stable]: 3,
};

function lowestStability(a?: Stability, b?: Stability): Stability | undefined {
  if (a === undefined) {
    return b;
  }
  if (b === undefined) {
    return a;
  }
  return stabilityPrecedence[a] < stabilityPrecedence[b] ? a : b;
}

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Runtime/JsonModel/Spec/Docs.cs
using Newtonsoft.Json;
using System.Collections.Generic;

namespace Amazon.JSII.JsonModel.Spec
{
    [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
    public sealed class Docs
    {
        public Docs
        (
            string? summary = null,
            string? remarks = null,
            string? deprecated = null,
            string? returns = null,
            Stability? stability = null,
            string? example = null,
            string? see = null,
            bool? subclassable = null,
            string? @default = null,
            IDictionary<string, string>? custom = null
        )
        {
            Summary = summary;
            Remarks = remarks;
            Deprecated = deprecated;
            Returns = returns;
            Stability = stability;
            Example = example;
            See = see;
            Subclassable = subclassable;
            Default = @default;
            Custom = custom ?? new Dictionary<string, string>();
        }

        [JsonProperty("summary", NullValueHandling = NullValueHandling.Ignore)]
        public string? Summary { get; }

        [JsonProperty("remarks", NullValueHandling = NullValueHandling.Ignore)]
        public string? Remarks { get; }

        [JsonProperty("deprecated", NullValueHandling = NullValueHandling.Ignore)]
        public string? Deprecated { get; }

        [JsonProperty("returns", NullValueHandling = NullValueHandling.Ignore)]
        public string? Returns { get; }

        [JsonProperty("stability", NullValueHandling = NullValueHandling.Ignore)]
        public Stability? Stability { get; }

        [JsonProperty("example", NullValueHandling = NullValueHandling.Ignore)]
        public string? Example { get; }

        [JsonProperty("see", NullValueHandling = NullValueHandling.Ignore)]
        public string? See { get; }

        [JsonProperty("subclassable", NullValueHandling = NullValueHandling.Ignore)]
        public bool? Subclassable { get; }

        [JsonProperty("default", NullValueHandling = NullValueHandling.Ignore)]
        public string? Default { get; }

        [JsonProperty("custom", NullValueHandling = NullValueHandling.Ignore)]
        public IDictionary<string, string> Custom { get; }


        public bool ShouldSerializeCustom() {
            return Custom.Count > 0;
        }
    }
}